lib/daemon.py

```diff
@@ -1,15 +1,44 @@
 import logging
 import os
 import sys
+import time
 
-from uaclient import http
+from uaclient import http, system
 from uaclient.config import UAConfig
 from uaclient.daemon import poll_for_pro_license, retry_auto_attach
 from uaclient.log import setup_journald_logging
 
 LOG = logging.getLogger("ubuntupro.daemon")
 
 
+# 10 seconds times 120 = 20 minutes
+WAIT_FOR_CLOUD_CONFIG_SLEEP_TIME = 10
+WAIT_FOR_CLOUD_CONFIG_POLL_TIMES = 120
+
+
+def _wait_for_cloud_config():
+    LOG.debug("waiting for cloud-config.service to finish")
+    for i in range(WAIT_FOR_CLOUD_CONFIG_POLL_TIMES + 1):
+        state = system.get_systemd_unit_active_state("cloud-config.service")
+        LOG.debug("cloud-config.service state: %r", state)
+        if state is not None and state == "activating":
+            if i < WAIT_FOR_CLOUD_CONFIG_POLL_TIMES:
+                LOG.debug(
+                    "cloud-config.service is activating. "
+                    "waiting to check again."
+                )
+                time.sleep(WAIT_FOR_CLOUD_CONFIG_SLEEP_TIME)
+            else:
+                LOG.warning(
+                    "cloud-config.service is still activating after "
+                    "20 minutes. continuing anyway"
+                )
+                return
+        else:
+            LOG.debug("cloud-config.service is not activating. continuing")
```

Comment:
should we log the state the service is actually in ? Maybe it is worth detecting an inactive vs failed service

Reply:
There is a separate log statement above on line 23 that includes the current state